File: app/lib/mealPlan.ts
import { apiFetch } from "./api";
import type { Recipe } from "./search";

export type MealPlanEntry = {
  day: string;
  slot: "breakfast" | "lunch" | "dinner" | "snack";
  /** Nécessaire pour demander la préparation de cette recette précise. */
  recipe_id: string;
  /** `instructions` est vide tant que la recette n'a pas été ouverte. */
  recipe: Recipe & { instructions?: string };
};

export type MealPlanView = {
  id: string;
  week_start: string;
  budget_target: number | null;
  estimated_cost: number | null;
  entries: MealPlanEntry[];
};

/** Monday of the current week, ISO date (YYYY-MM-DD). */
export function currentWeekStart(): string {
  const d = new Date();
  const day = (d.getDay() + 6) % 7; // 0 = Monday
  d.setDate(d.getDate() - day);
  return d.toISOString().slice(0, 10);
}

export async function generateMealPlan(
  weekStart: string,
  budget?: number
): Promise<MealPlanView> {
  return apiFetch("/meal-plan/generate", {
    method: "POST",
    body: JSON.stringify({ week_start: weekStart, budget: budget ?? null }),
  });
}

/**
 * Rédige la préparation d'une recette du plan, à sa première ouverture.
 *
 * Le plan ne les génère plus d'un bloc : elles pesaient ~74 % des tokens et
 * tronquaient la réponse. Le texte est conservé côté serveur, donc rouvrir la
 * même recette ne relance aucun appel.
 */
export async function fetchInstructions(recipeId: string): Promise<string> {
  const res = await apiFetch("/meal-plan/instructions", {
    method: "POST",
    body: JSON.stringify({ recipe_id: recipeId }),
  });
  return res.instructions as string;
}

File: api/app/models/meal_plan.py
from pydantic import BaseModel, Field

SLOTS = ("breakfast", "lunch", "dinner", "snack")


class Macros(BaseModel):
    kcal: float | None = None
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None


class PlanIngredient(BaseModel):
    name: str
    quantity: float | None = None
    unit: str | None = None


class PlanRecipeDraft(BaseModel):
    """Ce que le modèle produit à la génération du plan — sans la préparation.

    Les instructions représentaient ~74 % des tokens de sortie : 13 239 au
    total, contre un plafond de 16 000, ce qui tronquait le JSON par
    intermittence. Elles sont désormais rédigées à la demande, recette par
    recette (voir `recipe_instructions`).

    Les ingrédients, eux, restent générés ici : la liste de courses les agrège
    depuis la base pour toute la semaine, y compris les recettes jamais
    ouvertes.
    """

    title: str
    prep_time_min: int | None = None
    servings: int | None = None
    macros: Macros = Field(default_factory=Macros)
    ingredients: list[PlanIngredient]


class PlanRecipe(PlanRecipeDraft):
    """Ce que l'API renvoie. `instructions` est vide tant que l'utilisateur
    n'a pas ouvert la recette."""

    instructions: str = ""


class ScheduleEntry(BaseModel):
    day_offset: int = Field(description="0 = week_start, up to 6")
    slot: str = Field(description="breakfast | lunch | dinner | snack")
    recipe_index: int = Field(description="index into the recipes list")


class MealPlanGeneration(BaseModel):
    """Raw LLM output: deduped recipes + a schedule mapping into them."""

    recipes: list[PlanRecipeDraft]
    schedule: list[ScheduleEntry]
    estimated_cost: float | None = None


class RecipeSteps(BaseModel):
    """Préparation d'une recette, rédigée à la demande."""

    steps: list[str] = Field(description="4 à 7 étapes concrètes et cuisinables")


# ── API response shapes ──────────────────────────────────────────────
class MealPlanEntryView(BaseModel):
    day: str
    slot: str
    # Nécessaire au front pour demander la préparation de CETTE recette.
    recipe_id: str
    recipe: PlanRecipe


class MealPlanView(BaseModel):
    id: str
    week_start: str
    budget_target: float | None = None
    estimated_cost: float | None = None
    entries: list[MealPlanEntryView]


class GenerateRequest(BaseModel):
    week_start: str  # ISO date, the Monday of the target week
    budget: float | None = None


class InstructionsRequest(BaseModel):
    recipe_id: str


class InstructionsResponse(BaseModel):
    instructions: str
